Source/Core/S2Cell.js

```diff
@@ -0,0 +1,310 @@
+    throw new RuntimeError("S2 required BigInt support");
+  }
+  //>>includeStart('debug', pragmas.debug);
+  if (!defined(token)) {
+    throw new DeveloperError("token is required.");
+  }
+  if (!S2Cell.isValidToken(token)) {
+    throw new DeveloperError("token is invalid.");
+  }
+  //>>includeEnd('debug');
+
+  /**
+   * Gets or sets the cell token.
+   */
+  this._token = token;
+}
+
+/**
+ * Validates an S2 cell ID.
+ *
+ * @param {BigInt} [cellId] The S2CellId.
+ * @returns {Boolean} Returns true if the cell ID is valid, returns false otherwise.
+ */
+S2Cell.isValidId = function (cellId) {
+  //>>includeStart('debug', pragmas.debug);
+  Check.typeOf.bigint("cellId", cellId);
+  //>>includeEnd('debug');
+
+  if (cellId <= 0) return false;
+
+  // Check if face bits indicate a value <= 5.
+  if (cellId >> BigInt(S2PositionBits) > 5) return false;
+
+  // Check trailing 1 bit is in one of the even bit positions allowed for the 30 levels, using a bitmask.
+  var lowestSetBit = cellId & (~cellId + BigInt(1));
+  if (!(lowestSetBit & BigInt("0x1555555555555555"))) {
+    return false;
+  }
+
+  return true;
+};
+
+/**
+ * Validates an S2 cell token.
+ *
+ * @param {String} [token] The hexadecimal representation of an S2CellId. 0 must be represented using "X".
+ * @returns {Boolean} Returns true if the token is valid, returns false otherwise.
+ */
+S2Cell.isValidToken = function (token) {
+  //>>includeStart('debug', pragmas.debug);
+  Check.typeOf.string("token", token);
+  //>>includeEnd('debug');
+
+  var regex = new RegExp("^[0-9a-fA-F]{1,16}$");
+  if (!regex.test(token)) {
+    return false;
+  }
+
+  return S2Cell.isValidId(S2Cell.getIdFromToken(token));
+};
+
+/**
+ * Converts an S2 cell token to a 64-bit S2 cell ID.
+ *
+ * @param {String} [token] The hexadecimal representation of an S2CellId. 0 must be represented using "X". Expected to be a valid S2 token.
+ * @returns {BigInt} Returns the S2 cell ID.
+ */
+S2Cell.getIdFromToken = function (token) {
+  //>>includeStart('debug', pragmas.debug);
+  Check.typeOf.string("token", token);
+  //>>includeEnd('debug');
+
+  // 'X' is a special case of the S2Token, representing the 0 cell.
+  if (token === "X") {
+    return BigInt(0);
+  }
+
+  try {
+    return BigInt("0x" + token + "0".repeat(16 - token.length));
+  } catch (e) {
+    throw new RuntimeError(e);
+  }
+};
+
+/**
+ * Converts a 64-bit S2 cell ID to an S2 cell token.
+ *
+ * @param {BigInt} [cellId] The S2 cell ID.
+ * @returns {BigInt} Returns hexadecimal representation of an S2CellId.
+ */
+S2Cell.getTokenFromId = function (cellId) {
+  //>>includeStart('debug', pragmas.debug);
+  Check.typeOf.bigint("cellId", cellId);
+  //>>includeEnd('debug');
+
+  // 'X' is a special case of the S2Token, representing the 0 cell.
+  if (cellId === BigInt(0)) {
+    return "X";
+  }
+
+  try {
+    return cellId.toString(16).replace(/0*$/, "");
+  } catch (e) {
+    throw new RuntimeError(e);
+  }
```

Comment:
since we already asserted that cellId is a `BigInt`, it's safe to remove this try/catch block